runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -339,7 +339,8 @@ private void prepareStateBackend(K key) {
           // Key for state request is shipped already encoded as ByteString,
           // this is mostly a wrapping with ByteBuffer. We still follow the
           // usual key encoding procedure.
-          final ByteBuffer encodedKey = FlinkKeyUtils.encodeKey(key, keyCoder);
```

Comment:
Could we keep this? I think this nicely abstracts away any wrapping/array extraction.

Reply:
Since we know upfront that keys are ByteString, there is no need for the indirection and associated overhead. Perhaps we could also assign the key coder in the operator instead of passing it through the translation?

Reply:
I don't think the overhead is measurable. It's essentially a Noop.

> Perhaps we could also assign the key coder in the operator instead of passing it through the translation?

Is this part of the model, that we always expect ByteString for the key coder? If not, I'd leave it flexible because this is strictly a job-compile time micro optimization.

Reply:
It seems to be cast into the portability support classes and there is a comment here as well: https://github.com/apache/beam/blob/247e93ef7fbe87dbb1cd82a0ce6b2e6aa5254f7c/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java#L54